File: phpmailer-go/pop3.go
package phpmailer

import (
	"bufio"
	"fmt"
	"net"
	"strings"
	"time"
)

type POP3 struct {
	Host           string
	Port           int
	TimeoutSeconds int
	Connected      bool
}

func NewPOP3(host string, port int) *POP3 {
	return &POP3{
		Host:           host,
		Port:           port,
		TimeoutSeconds: 30,
		Connected:      false,
	}
}

func (p *POP3) Authorise(userName string, password string, debug bool) error {
	address := fmt.Sprintf("%s:%d", p.Host, p.Port)
	conn, err := net.DialTimeout("tcp", address, time.Duration(p.TimeoutSeconds)*time.Second)
	if err != nil {
		return wrapMailError(ErrorTransport, "pop3 connect failed", err)
	}
	defer conn.Close()

	_ = conn.SetReadDeadline(time.Now().Add(time.Duration(p.TimeoutSeconds) * time.Second))
	_ = conn.SetWriteDeadline(time.Now().Add(time.Duration(p.TimeoutSeconds) * time.Second))

	greeting, err := pop3ReadLine(conn)
	if err != nil {
		return err
	}
	if debug {
		fmt.Printf("POP3 greeting: %s\n", greeting)
	}
	if !strings.HasPrefix(greeting, "+OK") {
		return newMailError(ErrorProtocol, "invalid pop3 greeting: "+greeting)
	}

	if err := pop3Send(conn, "USER "+userName); err != nil {
		return err
	}
	userReply, err := pop3ReadLine(conn)
	if err != nil {
		return err
	}
	if !strings.HasPrefix(userReply, "+OK") {
		return newMailError(ErrorAuthentication, "pop3 USER failed: "+userReply)
	}

	if err := pop3Send(conn, "PASS "+password); err != nil {
		return err
	}
	passReply, err := pop3ReadLine(conn)
	if err != nil {
		return err
	}
	if !strings.HasPrefix(passReply, "+OK") {
		return newMailError(ErrorAuthentication, "pop3 PASS failed: "+passReply)
	}

	_ = pop3Send(conn, "QUIT")
	p.Connected = true
	return nil
}

func pop3Send(conn net.Conn, command string) error {
	_, err := conn.Write([]byte(command + "\r\n"))
	if err != nil {
		return wrapMailError(ErrorTransport, "pop3 write failed", err)
	}

	return nil
}

func pop3ReadLine(conn net.Conn) (string, error) {
	reader := bufio.NewReader(conn)
	line, err := reader.ReadString('\n')
	if err != nil {
		return "", wrapMailError(ErrorTransport, "pop3 read failed", err)
	}

	return strings.TrimRight(line, "\r\n"), nil
}

File: phpmailer-rs/src/pop3.rs
use std::io::{BufRead, BufReader, Write};
use std::net::TcpStream;
use std::time::Duration;

use crate::phpmailer::MailError;

#[derive(Debug, Clone)]
pub struct Pop3 {
    pub host: String,
    pub port: u16,
    pub timeout_seconds: u64,
    pub connected: bool,
}

impl Pop3 {
    pub fn new(host: String, port: u16) -> Self {
        Self {
            host: host.clone(),
            port,
            timeout_seconds: 30,
            connected: false,
        }
    }

    pub fn authorise(
        &mut self,
        user_name: String,
        password: String,
        debug: bool,
    ) -> Result<(), MailError> {
        let address = format!("{}:{}", self.host.clone(), self.port);
        let mut stream = TcpStream::connect(address.clone())
            .map_err(|err| MailError::transport(format!("POP3 connect failed for {address}: {err}")))?;

        stream
            .set_read_timeout(Some(Duration::from_secs(self.timeout_seconds)))
            .map_err(|err| MailError::transport(format!("POP3 read timeout failed: {err}")))?;

        stream
            .set_write_timeout(Some(Duration::from_secs(self.timeout_seconds)))
            .map_err(|err| MailError::transport(format!("POP3 write timeout failed: {err}")))?;

        let greeting = Self::read_line(&mut stream)?;
        if debug {
            println!("POP3 greeting: {}", greeting.clone());
        }
        if !greeting.starts_with("+OK") {
            return Err(MailError::protocol(format!(
                "Invalid POP3 greeting: {}",
                greeting.clone()
            )));
        }

        Self::send_command(&mut stream, format!("USER {}", user_name.clone()))?;
        let user_reply = Self::read_line(&mut stream)?;
        if !user_reply.starts_with("+OK") {
            return Err(MailError::authentication(format!(
                "POP3 USER failed for {}: {}",
                user_name.clone(),
                user_reply.clone()
            )));
        }

        Self::send_command(&mut stream, format!("PASS {}", password.clone()))?;
        let pass_reply = Self::read_line(&mut stream)?;
        if !pass_reply.starts_with("+OK") {
            return Err(MailError::authentication(format!(
                "POP3 PASS failed for {}: {}",
                user_name.clone(),
                pass_reply.clone()
            )));
        }

        Self::send_command(&mut stream, "QUIT".to_string())?;
        self.connected = true;

        Ok(())
    }

    fn send_command(stream: &mut TcpStream, command: String) -> Result<(), MailError> {
        stream
            .write_all(format!("{}\r\n", command.clone()).as_bytes())
            .map_err(|err| MailError::transport(format!("POP3 command write failed ({command}): {err}")))
    }

    fn read_line(stream: &mut TcpStream) -> Result<String, MailError> {
        let mut reader = BufReader::new(stream);
        let mut line = String::new();
        reader
            .read_line(&mut line)
            .map_err(|err| MailError::transport(format!("POP3 read line failed: {err}")))?;

        Ok(line.trim_end().to_string())
    }
}
